File: ICPC_Mirrors/Nitc_9.0/nwerc2019all/heightprofile/submissions/accepted/tobi.java
import java.io.*;
import java.util.*;

class pair {
	int first;
	double second;
	pair(int first, double second) {
		this.first = first;
		this.second = second;
	}
}

public class tobi {
	static long n, k;
	static long heights[] = new long[100005];
	static double g;
	static double diffs[] = new double[100005];
	static ArrayList<pair> candidates = new ArrayList<>();

	public static void main(String... args) throws Exception {
		Scanner scanner = new Scanner(System.in);
		n = scanner.nextLong();
		k = scanner.nextLong();
		n++;
		for (int i = 0; i < n; i++) heights[i] = scanner.nextLong();
		for (int i = 0; i < k; i++) {
			double best = 0;
			g = scanner.nextDouble();
			long wanted = (long)(g*10);
			for (int j = 0; j < n; j++) diffs[j] = heights[j] - wanted * j;
			candidates.clear();
			candidates.add(new pair(0, diffs[0]));
			for (int j = 1; j < n; j++) {
				int a = 0, o = candidates.size();
				while (a < o) {
					int m = (a + o) / 2;
					if (candidates.get(m).second <= diffs[j]) o = m;
					else a = m + 1;
				}
				if (a < candidates.size()) {
					int idx = candidates.get(a).first;
					while (idx > 0 && diffs[idx-1] <= diffs[j]) idx--;
					double streak = j - idx;
					best = Math.max(best, streak);
					if (idx > 0 && diffs[idx] != diffs[idx-1]) best = Math.max(best, streak + 1 - Math.min(1., (diffs[j] - diffs[idx-1]) / (diffs[idx] - diffs[idx-1])));
					if (j < n - 1 && diffs[j] != diffs[j+1]) best = Math.max(best, streak + Math.min(1., (diffs[j] - diffs[idx]) / (diffs[j] - diffs[j+1])));
				}
				if (diffs[j] < candidates.get(candidates.size() - 1).second) candidates.add(new pair(j, diffs[j]));
			}
			if (best == 0) System.out.println("impossible");
			else System.out.printf("%20.20f\n", best);
		}
	}
}

File: ICPC_Mirrors/Nitc_9.0/nwerc2019all/heightprofile/submissions/wrong_answer/tobi.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iomanip>
#define FOR(i,a,b) for (auto i=(a); i<(b); ++i)
using namespace std;

long long n, k, heights[100005];
double g, diffs[100005];
vector<pair<int, double>> candidates;

int main() {
	cin >> n >> k;
	n++;
	FOR(i, 0, n) cin >> heights[i];
	FOR(i, 0, k) {
		double best = 0;
		cin >> g;
		long long wanted = g*10;
		FOR(j, 0, n) diffs[j] = heights[j] - wanted * j;
		candidates.clear();
		candidates.push_back(make_pair(0, diffs[0]));
		FOR(j, 1, n) {
			int a = 0, o = candidates.size();
			while (a < o) {
				int m = (a + o) / 2;
				if (candidates[m].second <= diffs[j]) o = m;
				else a = m + 1;
			}
			if (a < candidates.size()) {
				int idx = candidates[a].first;
				while (idx > 0 && diffs[idx-1] <= diffs[j]) idx--;
				double streak = j - idx;
				best = max(best, streak);
				if (idx > 0) best = max(best, streak + 1 - (diffs[j] - diffs[idx-1]) / (diffs[idx] - diffs[idx-1]));
				if (j < n - 1) best = max(best, streak + (diffs[j] - diffs[idx]) / (diffs[j] - diffs[j+1]));
			}
			if (diffs[j] < candidates.back().second) candidates.push_back(make_pair(j, diffs[j]));
		}
		if (best == 0) cout << "impossible" << endl;
		else cout << fixed << setprecision(20) << best << endl;
	}
}
